File: go/internal/monorepo/workspace_manager.go
package monorepo

import (
	"encoding/json"
	"os"
	"path/filepath"

	"github.com/microsoft/beachball/internal/git"
	"gopkg.in/yaml.v3"
)

// WorkspaceManager identifies a monorepo/workspace manager.
type WorkspaceManager string

const (
	ManagerNpm   WorkspaceManager = "npm"
	ManagerYarn  WorkspaceManager = "yarn"
	ManagerPnpm  WorkspaceManager = "pnpm"
	ManagerLerna WorkspaceManager = "lerna"
	ManagerRush  WorkspaceManager = "rush"
)

// managerByFile maps config file names to their manager, in precedence order
// matching workspace-tools.
var managerByFile = map[string]WorkspaceManager{
	"lerna.json":           ManagerLerna,
	"rush.json":            ManagerRush,
	"yarn.lock":            ManagerYarn,
	"pnpm-workspace.yaml":  ManagerPnpm,
	"package-lock.json":    ManagerNpm,
}

// DetectWorkspaceManager determines the workspace manager for the given root directory.
func DetectWorkspaceManager(rootPath string) WorkspaceManager {
	// Check in precedence order (matching git.ManagerFiles)
	for _, file := range git.ManagerFiles {
		if _, err := os.Stat(filepath.Join(rootPath, file)); err == nil {
			return managerByFile[file]
		}
	}
	// Default to npm if package.json exists with workspaces
	return ManagerNpm
}

// GetWorkspacePatterns returns the workspace glob patterns (or literal paths for rush)
// for the detected manager at rootPath.
func GetWorkspacePatterns(rootPath string, manager WorkspaceManager) (patterns []string, literal bool) {
	switch manager {
	case ManagerPnpm:
		return getPnpmPatterns(rootPath), false
	case ManagerLerna:
		return getLernaPatterns(rootPath), false
	case ManagerRush:
		return getRushPaths(rootPath), true
	default: // npm, yarn
		return getNpmYarnPatterns(rootPath), false
	}
}

// getNpmYarnPatterns reads workspace patterns from package.json workspaces field.
func getNpmYarnPatterns(rootPath string) []string {
	data, err := os.ReadFile(filepath.Join(rootPath, "package.json"))
	if err != nil {
		return nil
	}

	// Try array format: "workspaces": ["packages/*"]
	var arrayFormat struct {
		Workspaces []string `json:"workspaces"`
	}
	if err := json.Unmarshal(data, &arrayFormat); err == nil && len(arrayFormat.Workspaces) > 0 {
		return arrayFormat.Workspaces
	}

	// Try object format: "workspaces": {"packages": ["packages/*"]}
	var objectFormat struct {
		Workspaces struct {
			Packages []string `json:"packages"`
		} `json:"workspaces"`
	}
	if err := json.Unmarshal(data, &objectFormat); err == nil && len(objectFormat.Workspaces.Packages) > 0 {
		return objectFormat.Workspaces.Packages
	}

	return nil
}

// getPnpmPatterns reads workspace patterns from pnpm-workspace.yaml.
func getPnpmPatterns(rootPath string) []string {
	data, err := os.ReadFile(filepath.Join(rootPath, "pnpm-workspace.yaml"))
	if err != nil {
		return nil
	}
	var config struct {
		Packages []string `yaml:"packages"`
	}
	if err := yaml.Unmarshal(data, &config); err != nil {
		return nil
	}
	return config.Packages
}

// getLernaPatterns reads workspace patterns from lerna.json.
// Falls back to npm/yarn/pnpm if lerna.json doesn't specify packages.
func getLernaPatterns(rootPath string) []string {
	data, err := os.ReadFile(filepath.Join(rootPath, "lerna.json"))
	if err != nil {
		return nil
	}
	var config struct {
		Packages []string `json:"packages"`
	}
	if err := json.Unmarshal(data, &config); err != nil {
		return nil
	}
	if len(config.Packages) > 0 {
		return config.Packages
	}

	// Lerna without packages: delegate to the actual package manager
	if _, err := os.Stat(filepath.Join(rootPath, "pnpm-workspace.yaml")); err == nil {
		return getPnpmPatterns(rootPath)
	}
	return getNpmYarnPatterns(rootPath)
}

// getRushPaths reads project paths from rush.json (literal paths, not globs).
func getRushPaths(rootPath string) []string {
	data, err := os.ReadFile(filepath.Join(rootPath, "rush.json"))
	if err != nil {
		return nil
	}
	var config struct {
		Projects []struct {
			ProjectFolder string `json:"projectFolder"`
		} `json:"projects"`
	}
	if err := json.Unmarshal(data, &config); err != nil {
		return nil
	}
	paths := make([]string, len(config.Projects))
	for i, p := range config.Projects {
		paths[i] = p.ProjectFolder
	}
	return paths
}

File: rust/src/monorepo/workspace_manager.rs
use std::path::Path;

use crate::git::commands::MANAGER_FILES;

/// Workspace/monorepo manager type.
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum WorkspaceManager {
    Npm,
    Yarn,
    Pnpm,
    Lerna,
    Rush,
}

/// Map from manager config file name to manager type.
fn manager_for_file(filename: &str) -> WorkspaceManager {
    match filename {
        "lerna.json" => WorkspaceManager::Lerna,
        "rush.json" => WorkspaceManager::Rush,
        "yarn.lock" => WorkspaceManager::Yarn,
        "pnpm-workspace.yaml" => WorkspaceManager::Pnpm,
        "package-lock.json" => WorkspaceManager::Npm,
        _ => WorkspaceManager::Npm,
    }
}

/// Detect the workspace manager by checking for config files in precedence order.
pub fn detect_workspace_manager(root: &str) -> WorkspaceManager {
    let root_path = Path::new(root);
    for file in MANAGER_FILES {
        if root_path.join(file).exists() {
            return manager_for_file(file);
        }
    }
    WorkspaceManager::Npm
}

/// Get workspace patterns for the detected manager.
/// Returns (patterns, is_literal) where is_literal=true means paths not globs (rush).
pub fn get_workspace_patterns(root: &str, manager: WorkspaceManager) -> (Vec<String>, bool) {
    match manager {
        WorkspaceManager::Pnpm => (get_pnpm_patterns(root), false),
        WorkspaceManager::Lerna => (get_lerna_patterns(root), false),
        WorkspaceManager::Rush => (get_rush_paths(root), true),
        _ => (get_npm_yarn_patterns(root), false),
    }
}

/// Read workspace patterns from package.json workspaces field (npm/yarn).
fn get_npm_yarn_patterns(root: &str) -> Vec<String> {
    let pkg_path = Path::new(root).join("package.json");
    let data = match std::fs::read_to_string(&pkg_path) {
        Ok(d) => d,
        Err(_) => return vec![],
    };

    // Try array format: "workspaces": ["packages/*"]
    #[derive(serde::Deserialize)]
    struct ArrayFormat {
        workspaces: Option<Vec<String>>,
    }
    if let Ok(parsed) = serde_json::from_str::<ArrayFormat>(&data) {
        if let Some(ws) = parsed.workspaces {
            if !ws.is_empty() {
                return ws;
            }
        }
    }

    // Try object format: "workspaces": {"packages": ["packages/*"]}
    #[derive(serde::Deserialize)]
    struct ObjectFormat {
        workspaces: Option<WorkspacesObject>,
    }
    #[derive(serde::Deserialize)]
    struct WorkspacesObject {
        packages: Option<Vec<String>>,
    }
    if let Ok(parsed) = serde_json::from_str::<ObjectFormat>(&data) {
        if let Some(ws) = parsed.workspaces {
            if let Some(pkgs) = ws.packages {
                if !pkgs.is_empty() {
                    return pkgs;
                }
            }
        }
    }

    vec![]
}

/// Read workspace patterns from pnpm-workspace.yaml.
fn get_pnpm_patterns(root: &str) -> Vec<String> {
    let yaml_path = Path::new(root).join("pnpm-workspace.yaml");
    let data = match std::fs::read_to_string(&yaml_path) {
        Ok(d) => d,
        Err(_) => return vec![],
    };

    #[derive(serde::Deserialize)]
    struct PnpmWorkspace {
        packages: Option<Vec<String>>,
    }

    match serde_yaml::from_str::<PnpmWorkspace>(&data) {
        Ok(config) => config.packages.unwrap_or_default(),
        Err(_) => vec![],
    }
}

/// Read workspace patterns from lerna.json. Falls back to npm/pnpm if absent.
fn get_lerna_patterns(root: &str) -> Vec<String> {
    let lerna_path = Path::new(root).join("lerna.json");
    if let Ok(data) = std::fs::read_to_string(&lerna_path) {
        #[derive(serde::Deserialize)]
        struct LernaConfig {
            packages: Option<Vec<String>>,
        }
        if let Ok(config) = serde_json::from_str::<LernaConfig>(&data) {
            if let Some(pkgs) = config.packages {
                if !pkgs.is_empty() {
                    return pkgs;
                }
            }
        }
    }

    // Lerna without packages: delegate to the actual package manager
    if Path::new(root).join("pnpm-workspace.yaml").exists() {
        return get_pnpm_patterns(root);
    }
    get_npm_yarn_patterns(root)
}

/// Read project paths from rush.json (literal paths, not globs).
fn get_rush_paths(root: &str) -> Vec<String> {
    let rush_path = Path::new(root).join("rush.json");
    let data = match std::fs::read_to_string(&rush_path) {
        Ok(d) => d,
        Err(_) => return vec![],
    };

    #[derive(serde::Deserialize)]
    struct RushConfig {
        projects: Option<Vec<RushProject>>,
    }
    #[derive(serde::Deserialize)]
    #[serde(rename_all = "camelCase")]
    struct RushProject {
        project_folder: String,
    }

    match serde_json::from_str::<RushConfig>(&data) {
        Ok(config) => config
            .projects
            .unwrap_or_default()
            .into_iter()
            .map(|p| p.project_folder)
            .collect(),
        Err(_) => vec![],
    }
}
